create proc/func/trigger, create/drop database, transaction statements are not being tracked by the extensions which utilize processutility hook

kritika-0601 opened the pull request:
Currently statements like create proc/func/trigger/database, drop database and transaction statements are not being tracked by the extensions which utilises the processutility hooks in Babelfish. Above issue happens because bbf_processutility doesn't call subsequent/previous processutility hook for above statements due to which these subsequent extension hooks are not being executed.
This commit resolves above issue by adding a custom processutility hook in engine to execute above Babelfish statements so that execution happens at last step of processutility hook chain which makes sure that subsequent extension hooks are executed.

Before:
processUtility_hooks()-> ... -> bbf_processUtility_hook() -> executes above statements -> execution completed/doesn't return to previous hook execution or doesn't execute subsequent hooks

After:
processUtility_hooks()-> ... -> bbf_processUtility_hook() -> ... -> processUtility_hooks()-> ... -> Standard_processUtility()-> executes above statements using custom hook -> returns to previous hook execution

Task: BABEL-4218

Since this PR includes calling the engine side function to extension side using hook, we do not require extra test scenarios to test it, it is already being covered in previous tests.
We have separate test PR for testing pg_stat_statements extension.

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV__PG_15_X' into pg_stat_statements-hook

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Transaction statements not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

surendravishnoi commented on src/backend/tcop/utility.c:
```diff
@@ -76,6 +76,7 @@
 
 /* Hook for plugins to get control in ProcessUtility() */
 ProcessUtility_hook_type ProcessUtility_hook = NULL;
+miscProcessUtility_hook_type miscProcessUtility_hook = NULL;
```
a better name would be bbfCustomProcessUtility_hook

surendravishnoi commented on src/include/tcop/utility.h:
```diff
@@ -16,6 +16,7 @@
 
 #include "tcop/cmdtag.h"
 #include "tcop/tcopprot.h"
+#include "parser/parse_node.h"
```
why this header is required? specifically for which struct

kritika-0601 replied:
It is required for ParseState.

surendravishnoi replied:
please use forward declaration

surendravishnoi commented on src/backend/tcop/utility.c:
```diff
@@ -594,6 +596,15 @@ standard_ProcessUtility(PlannedStmt *pstmt,
 	pstate->p_queryEnv = queryEnv;
 
+	if(miscProcessUtility_hook)
+		(*miscProcessUtility_hook)(pstate, pstmt, queryString, context, params, qc, &flag);
```
why not return true or false?

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

surendravishnoi commented on src/backend/tcop/utility.c:
```diff
@@ -559,6 +560,7 @@ standard_ProcessUtility(PlannedStmt *pstmt,
 	bool		isAtomicContext = (!(context == PROCESS_UTILITY_TOPLEVEL || context == PROCESS_UTILITY_QUERY_NONATOMIC) || IsTransactionBlock());
 	ParseState *pstate;
 	int			readonly_flags;
+	int flag = false;
```
not required

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV__PG_15_X' into pg_stat_statements-hook

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV__PG_15_X' into pg_stat_statements-hook

Commit:
retrigger tests

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

    Subsequent processutility hooks after bbf_processutilty are not being called.
    Created a hook for that and called it inside relevant
    ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
retrigger tests

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

        Subsequent processutility hooks after bbf_processutilty are not being called.
        Created a hook for that and called it inside relevant
        ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty are not being called.
Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

surendravishnoi commented on src/include/tcop/utility.h:
```diff
@@ -108,5 +108,12 @@ CreateCommandName(Node *parsetree)
+						  ParamListInfo params, QueryCompletion *qc);
+extern PGDLLIMPORT bbfCustomProcessUtility_hook_type bbfCustomProcessUtility_hook;
+
+void PLTsqlProcessTransaction(Node *parsetree,
```
why have we added this declaration here?

Commit:
Create proc/func/database etc not visible in pg_stat_statements view

    Subsequent processutility hooks after bbf_processutilty are not being called.
    Created a hook for that and called it inside relevant
    ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV__PG_15_X' into pg_stat_statements-hook
